packages/api/src/module/read/course-progress/course-progress.read-module.ts

```diff
@@ -47,15 +47,20 @@ export class CourseProgressReadModule {
 
   @OnEvent('LearningMaterialsTasks.TaskWasCompleted')
   async onTaskWasCompleted(event: ApplicationEvent<TaskWasCompleted>) {
-    await this.prismaService.courseProgress.update({
+    await this.prismaService.courseProgress.upsert({
```

Comment:
You should do something similar while handling all events here. 
If we don't force ordering (TaskWasCompleted before LearningMaterialUrlWasGenerated) etc. 
everything may cause create or update (upsert). Because if TaskWasCompleted created CourseProgress entry what about `create` in LearningMaterialsUrlWasGenerated? It's 100% fail.

You can also remove previously added while handling TaskWasUncompleted - read model it's not place for domain rule: 
```ts
if (!courseProgress || courseProgress.learningMaterialsCompletedTasks === 0) {
      return;
    }
```
Try to do it and reduce code duplication :)